### PR TITLE
[Android] Adding Fluent Icon support in Adaptive Cards

### PR DESCRIPTION
**Pull request by singhmanp0707**
# Description

This PR contains the following:
1. Adding Fluent Icons support in Adaptive Cards. A new fluent icon element is introduced.
2. Rendering the fluent icons in actions by using the existing iconUrl property.

Spec doc - https://microsoft-my.sharepoint-df.com/:w:/p/dclaux/EXXwg2VEWQNImweX246LoWYBv6C8KK-DbgneUG5Wz5p2Qw?e=KZWrjd

CDN directory URL - https://[res-1.cdn.office.net/assets/fluentui-react-icons/2.0.226/directory.json](https://res-1.cdn.office.net/assets/fluentui-react-icons/2.0.226/directory.json)

Sample Icon CDN URL - https://[res-1.cdn.office.net/assets/fluentui-react-icons/2.0.226/Calendar/Calendar24Filled.json](https://res-1.cdn.office.net/assets/fluentui-react-icons/2.0.226/Calendar/Calendar24Filled.json)

Following is spec for this new item:

| Property | ValueType | 
| -------- | -------- | 
| type   | String = "Icon"     | 
| name    | String     | 
| size    | String = one of these values: “ xxSmall ” (12px) “ xSmall ” (16px) “Small” (20px) “Standard” (24px) “Medium” (28px) “Large” (32px) “ xLarge ” (40px) “ xxLarge ” (48px)     | 
| style   | String = “regular” or “filled” | 
| color    | String = “default ”, “dark ”, “light ”, “accent ”, “good ”, “warning” or “attention” These are the same as text colors.     | 
| selectAction    | Action. All action types except Action.ShowCard     | 

# Change in Action element

The existing Action.iconUrl property will accept a new format, namely:

icon:\<icon name>\[,regular | filled] 

is the name of one of the available icons (see the “Icon directory” section below)

The style specifier is optional. If not specified, it defaults to “regular”

It is not possible for the developer to control the size or color of icons in buttons. The product (Teams, Outlook) fully controls those aspects.

# Validation

### Without RTL:

FluentIconTeams

### With RTL:

FluentIconTeams_RTL

### Without RTL:

FluentIcon

### With RTL:

FluentIconRTL

### CODE REVIEW

**Comment by ardlank** [comment]
Can you separate out Android and iOS work into separate PRs? Also change PR from Draft to Open

**Comment by singhmanp0707** [comment]
> Can you separate out Android and iOS work into separate PRs? Also change PR from Draft to Open

I'll wait for the iOS PR to be checked in. After that, I'll take a pull from main and publish the PR.

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/FluentIconImageLoaderAsync.kt [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * Responsible for rendering the Fluent Icon element
+ **/
+open class FluentIconImageLoaderAsync(
+    val renderedCard: RenderedAdaptiveCard,
+    val iconSize: Long,
+    private val iconColor: String,
+    view: View
+) : AsyncTask<String, Void, HttpRequestResult<String>>() {
```
Comment (karthikgopal36) [inline source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/FluentIconImageLoaderAsync.kt:30]: Can we use `Coroutine`? if it is a big change, will take care of it in an another PR.

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/FluentIconImageLoaderAsync.kt [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        return if (args.isEmpty()) {
+            null
+        } else {
+            fetchIcon(args[0])
+        }
+    }
+
+    override fun onPostExecute(result: HttpRequestResult<String>?) {
+        val view = viewReference.get()
+        if (view != null) {
```
Comment (karthikgopal36) [inline source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/FluentIconImageLoaderAsync.kt:43]: view?.let{}

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/FluentIconImageLoaderAsync.kt (hunk 1 above)
Comment (singhmanp0707) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/FluentIconImageLoaderAsync.kt:30]: I used the existing infra and dependencies in the SDK. It'll be big refactoring. This can be taken up separately in another PR.

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/FluentIconImageLoaderAsync.kt (hunk 2 above)
Comment (singhmanp0707) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/FluentIconImageLoaderAsync.kt]: done

File: source/shared/cpp/ObjectModel/Icon.h [hunk 3]
```diff
@@ -25,13 +25,13 @@ class Icon : public BaseCardElement
     ~Icon() = default;
 
     Json::Value SerializeToJsonValue() const override;
-
```
Comment (prpatwa) [inline source/shared/cpp/ObjectModel/Icon.h:28]: nit unnecessary changes, can be removed.

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/ActionElementRendererFluentIconImageLoaderAsync.kt [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                    drawables[1] = flippedDrawable
+                } else {
+                    drawables[0] = flippedDrawable
+                    view.compoundDrawablePadding = Util.dpToPixels(view.context, padding.toFloat())
+                }
+
+                view.setCompoundDrawablesRelativeWithIntrinsicBounds(
+                    drawables[0],
+                    drawables[1],
+                    drawables[2],
```
Comment (prpatwa) [inline source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/ActionElementRendererFluentIconImageLoaderAsync.kt:47]: it should be wrapped in drawable size check as well to avoid OutOfBounds exception

**Review by prpatwa — APPROVED** [review]

**Review by wyAbhishek — APPROVED** [review]
LGTM

File: source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/ActionElementRendererFluentIconImageLoaderAsync.kt (hunk 4 above)
Comment (singhmanp0707) [reply source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/ActionElementRendererFluentIconImageLoaderAsync.kt:47]: drawable size check is not required. 

view.compoundDrawablesRelative returns drawables for the start, top, end, and bottom borders.

File: source/shared/cpp/ObjectModel/Icon.h (hunk 3 above)
Comment (singhmanp0707) [reply source/shared/cpp/ObjectModel/Icon.h:28]: Tried reverting this earlier. Android studio is applying its formatting on the file, hence these empty spaces are getting added. Leaving it as it is.

**Review by ardlank — APPROVED** [review]

**Review by karthikgopal36 — APPROVED** [review]
